### ISSUE TITLE
metrics calculation with callbacks when using fit_generator with generators for both training and validation

### ISSUE THREAD

**Issue by genghis88** [keras-team/keras#7462]
If generator is used for the parameter validation_data when calling fit generator, how do you calculate custom metrics at the end of each epoch? Just wanted to try and understand how others are going about this problem.

Here's the code I use:

``model.fit_generator(generate_data_from_file('data/0.1-percent/training-data.tsv', binarizer, batch_size),
        validation_data=generate_data_from_file('data/0.1-percent/validation-data.tsv', binarizer, batch_size),
        steps_per_epoch=math.ceil(1.0 * 109527 / batch_size),
        validation_steps=math.ceil(1.0 * 13692 / batch_size),
        epochs=10,
        verbose=1,
        class_weight=class_weights,
        max_queue_size=2)``

**Comment by fabriciorsf** [comment]
I did this:

```
history = CumulativeHistory()
callbacks = [history]
from keras import backend as K
if K.backend() == 'tensorflow':
  board = keras.callbacks.TensorBoard(log_dir=f"{self.prefix_folder_logs}{time()}",
                                    histogram_freq=1, write_graph=True, write_images=True)
  callbacks.append(board)
metric_to_compare = 'val_euclidean_distance'
print("Begin of training model...")
for i in range(MAX_NUM_EPOCHS):
  model.fit_generator(self.train_inputs, steps_per_epoch=self.train_inputs.steps_per_epoch(),
                      validation_data=test_input_sequence, validation_steps=steps_test,
                      max_queue_size=self.train_inputs.workers, epochs=i+1, initial_epoch=i,
                      workers=self.train_inputs.workers, use_multiprocessing=True,
                      callbacks = callbacks)
  try:
    metrics_diff = history.history[metric_to_compare][i] - min(history.history[metric_to_compare][:i])
  except:
    metrics_diff = -1
  if metrics_diff < 0:
    self._save_models(i)
    self.data_processor = None  # Empty memory
    best_epoch = i
    num_worse_epochs = 0
  elif metrics_diff > 0:
    num_worse_epochs += 1
    if num_worse_epochs >= PATIENCE:
      print("Ran out of patience. Stopping training.")
      break
print("End of training model.")
```

**Comment by Times1035** [comment]
I have some questions about it.
I am now  using fit_generator with generators for both training and validation.
I've wrote a class named My_callback which is inherited from keras.callbacks.Callback, in the class,  I define several metrics like loss,val_loss,acc,val_acc. 
But when run the model.fit_generator code, only loss and acc can be computed, there will be a error printed  KeyError:'val_loss'

Because my data is so large, so I split the data into 32 files, each file has 10000 samples.

**my simplified code is:**

def batch_iter(xpath,ypath,batch_size,num_epochs,shuffle=True):
    num_file = len(xpath)
    for epoch in range(num_epochs):
        for i in range(num_file):
            X = np.load(xpath[i])[:,:,:,np.newaxis]
            Y = np.load(ypath[i])
            data_size = len(X)
            num_batches_per_file = int((len(X)-1)/batch_size)+1
            if shuffle:
                shuffle_indices = np.random.permutation(np.arange(data_size))
                shuffle_X = X[shuffle_indices]
                shuffle_Y = Y[shuffle_indices]
            else:
                shuffled_X = X
                shuffle_Y = Y
            for batch_num in range(num_batches_per_file):
                start_index = batch_num * batch_size
                end_index = min((batch_num+1)*batch_size,data_size)
                x_batch = shuffle_X[start_index:end_index]
                y_batch = shuffle_Y[start_index:end_index]
                yield (x_batch,y_batch)

def top_2_accuracy(y_true,y_pred):
    return keras.metrics.sparse_top_k_categorical_accuracy(y_true,y_pred,k=2)

def top_3_accuracy(y_true,y_pred):
    return keras.metrics.sparse_top_k_categorical_accuracy(y_true,y_pred,k=3)

def cnn_model():
    model = Sequential()
    model.add(Conv2D(filters=32,kernel_size=(3,100),padding='same',input_shape=X_train.shape[1:],))
    model.add(Activation('relu'))
    model.add(MaxPool2D(pool_size=(25,1),strides=None))
    model.add(Flatten())
    model.add(Dense(units=128,activation='relu'))
    model.add(Dropout(0.25))
    model.add(Dense(units=Y_trian.shape[1],activation='sigmoid'))
    model.summary
    opt = optimizers.Adam()
    model.compile(loss='binary_crossentropy',optimizer=opt,metrics=['accuracy',top_2_accuracy,top_3_accuracy,'sparse_top_k_categorical_accuracy'])
    return model

class My_Callback(keras.callbacks.Callback):
    def __init__(self,batch_interval):
        self.metrics = metrics
        self.batch_interval = batch_interval
        self.loss = []
        self.val_loss = []

    def on_train_begin(self,log={}):
        self.loss = []
        self.val_loss = []
        self.batch_number = 0
        self.epoch_number = 0

    def on_epoch_begin(self,logs={}):
        return

    def on_batch_end(self,batch,logs={}):
        self['batch_number'] +=1
        if self.batch_number % batch_interval == 0:
            self.epoch_number +=1        
            self.loss.append(logs['loss'])
            self.loss.append(logs['val_loss'])
        return

x_path_train,y_path_train,x_path_val,y_path_val,x_path_test,y_path_test = train_val_test_split(x_path,y_path,1,1,1)
model = cnn_model()

history = My_Callback(batch_interval=10000*len(x_path_train)/16)
model.fit_generator(batch_iter(x_path_train,y_path_train,16,10),
                    steps_per_epoch = 10000*10*len(x_path_train)/16,
                    validation_data = batch_iter(x_path_val,y_path_val,16,10,shuffle=False),
                    validation_stes = 10000*10*len(x_path_val)/16,
                    callbacks = [history])
